core/src/mindustry/world/modules/ItemModule.java

```diff
@@ -66,11 +66,28 @@ public int total(){
         return total;
     }
 
+    // Making the take() loop persistent so it does not return the same item twice in a row unless there is nothing else to return.
+    protected int takeRotation;
+
     public Item take(){
-        for(int i = 0; i < items.length; i++){
+    	// 0-to-length loop broken in two parts. First resume the loop where it previously left off.
+        for(int i = takeRotation; i < items.length; i++){
+            if(items[i] > 0){
+                items[i]--;
+                total--;
+    		// save the next position so the next call to take() can resume from there.
+                takeRotation = (i + 1) % items.length;
+                return content.item(i);
+            }
+        }
```

Comment:
This method can be written as a single loop.

```java
for(int i = 0; i < items.length; i++){
    int index = (i + takeRotation) % items.length;
    if(items[index] > 0){
        items[index] --;
        total --;
        takeRotation = (index + 1) % items.length;
        return content.item(i);
    }
}
```

Reply:
Sure, it can, though modulo remains an extremely expensive instruction on modern hardware (from 15-25 cycles for IDIV/r8 to 50-70 cycles for /r64 vs less than one cycle effective cost for most other 1-5 cycles ALU instructions due to pipelined superscalar execution) and depending on how the java compiler re-factors this, you can conceivably spend more CPU time waiting for IDIV to produce the remainder the rest of the loop depends on than doing useful work in tight loops like that.

As a mostly low-level/bare-metal ASM/C/C++/VHDL programmer, putting an exceptionally expensive operation in an otherwise trivial loop's control path when it can so easily be avoided bothers me.

Hope you found this bit of low-level hardware perspective enlightening.

Reply:
I am aware that modulo is generally considered expensive, but in this case there isn't much of a benefit to splitting things up into two loops. Adding an extra loop adds more code and makes this method harder to read and modify.

You also have to consider the fact that this is extremely premature optimization. `take()` is a method that is called only once per frame for every input bridge and unloader. Considering that most active games have few amounts of these blocks being used, it really doesn't matter which implementation is slower, because the difference is only going to be a few nanoseconds. I would rather have more readable code than something that's ugly and imperceptibly faster.

That being said, I tried benchmarking both versions to see how significant the speed difference was, and the performance difference (with random items) for these two is about ~12 ns, with the non-modulo one taking 30 and one taking 42. Rewriting the mod implementation to mostly use subtraction instead of modulo actually made it run faster than yours, with 21 ns taken per invocation. This benchmark probably isn't too accurate, but when dealing with nanoseconds, it really doesn't matter.

Code:
```java
            for(int i = 0; i < items.length; i++){
                int index = (i + takeRotation);
                if(index >= items.length) index -= items.length; //conditional instead of mod
                if(items[index] > 0){
                    items[index] --;
                    total --;
                    takeRotation = index + 1;
                    return content.item(index % items.length); //called rarely so it doesn't really matter here, according to benchmarks
                }
            }

            return null;
```

Reply:
Guessing this is speculative execution at work: branch prediction may not like my front loop having a pseudo-random starting point so your more predictable loop ends up cheaper overall despite having an extra test.

As for being imperceptibly faster, imperceptibly faster bits of code add up over time.